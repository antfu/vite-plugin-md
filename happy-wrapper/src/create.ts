import { identity } from 'fp-ts/lib/function'
import { Text, Window } from 'happy-dom'
import { dasherize } from 'native-dash'
import { HappyMishap } from './errors'
import type { Container, HTML } from './happy-types'
import { isElement, isElementLike, isTextNodeLike } from './type-guards'
import { clone, solveForNodeType, toHtml } from './utils'
import type { Document, Fragment, IElement, IText } from './index'

/**
 * Converts an HTML string into a Happy DOM document tree
 */
export function createDocument(body: string, head?: string): Document {
  const window = new Window()
  const document = window.document
  document.body.innerHTML = body
  if (head)
    document.head.innerHTML = head
  return document
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type FragmentFrom<_T extends Container | 'html'> = Fragment

export function createFragment<C extends Container | HTML>(content?: C): FragmentFrom<C extends string ? 'html' : Fragment> {
  const window = new Window()
  const document = window.document
  const fragment = document.createDocumentFragment() as Fragment
  if (content)
    fragment.append(clone(content))

  return fragment as FragmentFrom<C extends string ? 'html' : Fragment>
}

/**
 * Creates a DOM Node which will be either an `IElement` or `IText` node
 * based on the content passed in.
 */
export const createNode = (node: Container | string): IElement | IText => {
  const frag = createFragment(node)
  if (isElementLike(frag))
    return frag.firstElementChild as IElement
  else if (isTextNodeLike(frag))
    return frag.firstChild as IText
  else
    throw new HappyMishap('call to createNode() couldn\'t be converted to IElement or IText node', { name: 'createNode()', inspect: node })
}

/**
 * Creates a IText Node
 */
export function createTextNode(text?: string): IText {
  if (!text) {
    console.warn('An empty string was passed into createTextNode(); will be ignored but probably a mistake')
    return new Text('')
  }

  const frag = createFragment(text)
  if (isTextNodeLike(frag))
    return frag.firstChild as unknown as IText
  else
    throw new HappyMishap(`The HTML passed in cannot be converted to a single text node: "${text}".`, { name: 'createFragment(text)', inspect: frag })
}

/**
 * Creates an element node and can preserve parent relationship if known
 */
export const createElement = (el: Container | HTML, parent?: IElement): IElement => solveForNodeType()
  .outputType<IElement>()
  .solver({
    node: (n) => {
      if (isElement(n))
        return createElement(n) as IElement
      else
        throw new HappyMishap('can\'t create an IElement from an INode node because it doesn\'t have a tagName property', { inspect: n })
    },
    html: (h) => {
      const frag = createFragment(h)
      if (isElementLike(frag)) {
        if (parent) {
          parent.append(frag.firstElementChild)
          return parent?.lastElementChild
        }

        return frag.firstElementChild
      }
      else { throw new HappyMishap('The HTML passed into createElement() is not convertible to a IElement node!', { name: 'createElement(html)', inspect: frag }) }
    },
    element: identity,
    text: (t) => {
      throw new HappyMishap('An IElement can not be created from a IText node because element\'s require a wrapping tag name!', { name: 'createElement(text)', inspect: t })
    },
    fragment: (f) => {
      if (isElement(f.firstElementChild))
        return f.firstElementChild as IElement

      else
        throw new HappyMishap(`Unable to create a IElement node from: \n\n${toHtml(f)}`, { name: 'createElement()' })
    },
    document: (d) => {
      if (isElementLike(d)) {
        if (parent)
          throw new HappyMishap('A Document and a parent IElement were passed into createElement. This is not a valid combination!')

        return d.firstElementChild
      }

      else { throw new HappyMishap('Can not create an Element from passed in Document', { name: 'createElement(document)', inspect: d }) }
    },
  })(el)

export interface CssVariable {
  prop: string
  value: string | number | boolean
}

export type ClassDefn = Record<string, string | boolean | number>
export type MultiClassDefn = [selector: string, keyValues: ClassDefn][]

export interface ClassApi {
  /** add a child selector */
  addChild: (selector: string, defn: ClassDefn) => ClassApi
  /** add CSS prop/values */
  addProps: (defn: ClassDefn) => ClassApi
}

export interface InlineStyle {
  /** add a single CSS variable (at a time); the CSS scope will ':root' unless specified */
  addCssVariable: (prop: string, value: string | number | boolean, scope?: string) => InlineStyle
  addClassDefinition: (selection: string, cb: ((api: ClassApi) => void)) => InlineStyle
  addCssVariables: (dictionary: Record<string, string>, scope?: string) => InlineStyle
  convertToVueStyleBlock: (lang: 'css' | 'scss', scoped: boolean) => InlineStyle

  finish: () => IElement
}

const renderClasses = (klasses: MultiClassDefn) => {
  return klasses.map(
    ([selector, defn]) => `\n\n  ${selector} {\n${Object.keys(defn).map(
      p => `    ${dasherize(p)}: ${defn[p]};`,
    ).join('\n')}\n  }`).join('\n')
}

/**
 * Creates a new `<style>...</style>` node and provides a simple API surface to allow
 * populating the contents with inline CSS content
 */
export const createInlineStyle = <T extends string = 'text/css'>(type: T = 'text/css' as T) => {
  const cssVariables: Record<string, CssVariable[]> = {}
  const cssClasses: MultiClassDefn = []
  let isVueBlock = false
  let isScoped = true
  let vueLang: 'css' | 'scss' = 'css'

  const api: InlineStyle = {
    addCssVariable(prop: string, value: string | number | boolean, scope = ':root') {
      if (!(scope in cssVariables))
        cssVariables[scope] = []
      cssVariables[scope].push({ prop: prop.replace(/^--/, ''), value })

      return api
    },
    addClassDefinition(selector, cb) {
      const classApi: ClassApi = {
        addChild: (child, defn) => {
          const childSelector = `${selector} ${child}`
          cssClasses.push([childSelector, defn])
          return classApi
        },
        addProps: (defn) => {
          cssClasses.push([selector, defn])
          return classApi
        },
      }
      cb(classApi)
      return api
    },
    addCssVariables(dictionary: Record<string, string | number | boolean>, scope = ':root') {
      Object.keys(dictionary).forEach(p => api.addCssVariable(p, dictionary[p], scope))

      return api
    },
    convertToVueStyleBlock(lang, scoped = true) {
      vueLang = lang
      isVueBlock = true
      isScoped = scoped

      return api
    },

    finish() {
      const setVariable = (scope: string, defn: Record<string, any>) => `${scope} {\n${Object.keys(defn).map(prop => `    --${defn[prop].prop}: ${defn[prop].value}${String(defn.prop).endsWith(';') ? '' : ';'}`).join('\n')}\n  }`

      let text = ''
      // variables
      Object.keys(cssVariables).forEach(
        (v) => {
          text = `${text}${setVariable(v, cssVariables[v])}\n`
        },
      )
      // classes
      text = `${text}${renderClasses(cssClasses)}`

      return isVueBlock
        ? createElement(`<style lang="${vueLang}"${isScoped ? ' scoped' : ''}>\n${text}\n</style>`)
        : createElement(`<style type="${type}">\n${text}\n</style>`)
    },
  }

  return api
}
