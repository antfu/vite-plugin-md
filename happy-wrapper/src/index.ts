export { Document, DocumentFragment as Fragment, IElement, IText } from 'happy-dom'

export * from './attributes'
export * from './create'
export * from './diagnostics'
export * from './errors'
export * from './happy-types'
export * from './nodes'
export * from './select'
export * from './type-guards'
export * from './utils'
