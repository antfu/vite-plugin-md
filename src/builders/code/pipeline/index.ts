export * from './addLanguage'
export * from './addClipboard'
export * from './convertBlocksToDomNodes'
export * from './defaultBlocks'
export * from './extractMarkdownItTokens'
export * from './expandCodeBlockVariables'
export * from './inlineStyles'
export * from './highlightLines'
export * from './renderHtml'
export * from './updateLineNumbers'
export * from './useHighlighter'
export * from './userRules'
export * from './updatePreWrapper'
export * from './updateFrontmatterWithCodeBlock'
export * from './updateCodeBlockWrapper'
