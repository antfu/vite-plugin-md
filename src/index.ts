import type { Plugin } from 'vite'
import { createFilter } from '@rollup/pluginutils'
import { createMarkdown } from './markdown'
import { resolveOptions } from './options'
import type { Options } from './types'

function VitePluginMarkdown(userOptions: Options = {}): Plugin {
  const options = resolveOptions(userOptions)
  const markdownToVue = createMarkdown(options)

  const filter = createFilter(
    userOptions.include || /\.md$/,
    userOptions.exclude,
  )

  return {
    name: 'vite-plugin-md',
    enforce: 'pre',
    transform(raw, id) {
      if (!filter(id))
        return
      try {
        return markdownToVue(id, raw)
      }
      catch (e: any) {
        this.error(e)
      }
    },
    async handleHotUpdate(ctx) {
      if (!filter(ctx.file))
        return

      const defaultRead = ctx.read
      ctx.read = async function() {
        return markdownToVue(ctx.file, await defaultRead())
      }
    },
  }
}

export default VitePluginMarkdown
