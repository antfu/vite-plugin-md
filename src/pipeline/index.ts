export * from './applyMarkdownItOptions'
export * from './baseStyling'
export * from './composeSfcBlocks'
export * from './createParser'
export * from './convertHandlerToTask'
export * from './convertToDom'
export * from './escapeCodeTagInterpolation'
export * from './extractBlocks'
export * from './extractFrontmatter'
export * from './finalize'
export * from './pipelineUtilityFunctions'
export * from './frontmatterPreprocess'
export * from './loadMarkdownItPlugins'
export * from './parse'
export * from './repairFrontmatterLinks'
export * from './transformsBefore'
export * from './wrapHtml'
