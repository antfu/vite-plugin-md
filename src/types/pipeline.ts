import type MarkdownIt from 'markdown-it'
import type * as TE from 'fp-ts/TaskEither'
import type { UserConfig } from 'vite'
import type { Either } from 'fp-ts/lib/Either'
import type { Task } from 'fp-ts/lib/Task'
import type { EnumValues, Frontmatter, MetaProperty, ResolvedOptions } from './core'

export enum PipelineStage {
  /**
   * Initialized with incoming filename, config, options,
   * available events (core and provided by builders).
   */
  initialize = 'initialize',

  /**
   * All frontmatter has been extracted from default values and page values
   * but no mapping has been done yet.
   *
   * Note: this is the event hook which the included `meta` builder connects
   * to and it in turn _provides_ a `metaMapped` hook.
   */
  metaExtracted = 'metaExtracted',

  /**
   * The **MarkdownIt** parser is initialized, all builders
   * connecting at this point will receive a valid `md` parser
   * object so that they can participate in MD-to-HTML parsing.
   */
  parser = 'parser',
  /**
   * The **MarkdownIt** parser is initialized and all builders
   * have been able to apply their customizations to it.
   */
  parsed = 'parsed',

  /**
   * SFC blocks (template, script, and an array of customBlocks) are ready for
   * builders to inspect/mutate/etc.
   */
  sfcBlocksExtracted = 'sfcBlocksExtracted',

  /**
   * All mutations of page are complete; builders can hook into this stage but
   * will _not_ be able to mutate at this stage.
   */
  closeout = 'closeout',
}

export type IPipelineStage = EnumValues<PipelineStage>

export interface RulesUse {
  ruleName: string
  usage: 'adds' | 'patches' | 'modifies'
  description?: string
}

export type PipelineInitializer = (i?: Pipeline<PipelineStage.initialize>) => Pipeline<PipelineStage.initialize>

export interface BuilderRegistration<O extends BuilderOptions, S extends IPipelineStage> {
  name: string
  description?: string
  /** The lifecycle event/hook which this builder will respond to */
  lifecycle: S
  /**
   * The builder's handler function which receives the _payload_ for the
   * event lifecycle hook configured and then is allowed to mutate these
   * properties and pass back a similarly structured object to continue
   * on in that pipeline stage.
   */
  handler: BuilderHandler<O, S>

  /**
   * The options _specific_ to the builder
   */
  options: O

  /**
   * This isn't strictly required, but it is nice to express which rules you have used
   * modified, or added from the MarkdownIt parser.
   *
   * Note: builders should try to avoid mutating core rules; if they need a modification
   * for their purposes consider _monkey patching_ the rule so that downstream rules
   * have a better understanding of current rule state.
   */
  parserRules?: RulesUse[]

  /**
   * If this plugin needs to modify the configuration in some way at initialization
   * it can add a function here to do that. In most cases, the builder can simply
   * wait for their event hook to be called (at which point they will get the configuration
   * passed to them).
   */
  initializer?: BuilderHandler<O, PipelineStage.initialize>
}

/** container of events organized by PipelineStage */
export type BuilderStruct<T extends string> = Record<PipelineStage | T, any>

export interface PipelineProperties {
  fileName: string
  /** the raw content in the file being processed */
  content: string
  /** the `vite-plugin-md` options */
  options: ResolvedOptions
  /** the Vite config */
  viteConfig: UserConfig

  /** the frontmatter metadata */
  frontmatter: Frontmatter
  /**
   * Meta properties that will be put into the HEAD section
   */
  meta: MetaProperty[]
  /**
   * Meta properties which are to be added to the VueJS router's "meta" attribute
   * for this page's route
   */
  routeMeta: Record<string, any>
  /**
   * Non-meta tags that will be put into the HEAD section of the page
   */
  head: Record<string, any>
  excerpt?: string
  /**
   * The markdown content (after extracting frontmatter)
   */
  md: string

  /** the **MarkdownIT** parser instance */
  parser: MarkdownIt

  /**
   * the HTML produced from MD content (and using parser rules passed in)
   */
  html: string

  /**
   * If any code blocks were found on the page then their languages will be represented
   * here.
   */
  fencedLanguages: Set<string>

  /**
   * all hoisted scripts
   */
  hoistedScripts: string[]

  /** the SFC's template block (aka, html content) */
  templateBlock: string
  /** the `<script [setup] ...>` block */
  scriptBlock: string
  /** any other top-level SFC blocks besides "template" and "script" */
  customBlocks: string[]

  /** the finalized component in string form */
  component: string
}

export type InitializedOmissions = 'md'
| 'fencedLanguages'
| 'frontmatter'
| 'head'
| 'meta'
| 'routeMeta'
| 'excerpt'
| 'html'
| 'hoistedScripts'
| 'templateBlock'
| 'parser'
| 'scriptBlock'
| 'customBlocks'
| 'component'

/** after extracting metadata */
export type MetaOmissions =
  | 'fencedLanguages'
  | 'parser'
  | 'html'
  | 'hoistedScripts'
  | 'templateBlock'
  | 'scriptBlock'
  | 'customBlocks'
  | 'component'

/** after providing the markdown-it parser */
export type ParserOmissions =
  | 'fencedLanguages'
  | 'html'
  | 'hoistedScripts'
  | 'templateBlock'
  | 'scriptBlock'
  | 'customBlocks'
  | 'component'

/** after parsing to raw HTML using markdown-it */
export type ParsedOmissions =
  | 'hoistedScripts'
  | 'templateBlock'
  | 'scriptBlock'
  | 'customBlocks'
  | 'component'

export type SfcBlockOmissions = 'component'

// export type PipelineAvail<S extends IPipelineStage> = S extends 'initialize'
//   ? Omit<PipelineProperties, InitializedOmissions>
//   : S extends 'metaExtracted' ? Omit<PipelineProperties, MetaOmissions>
//     : S extends 'parser' ? Omit<PipelineProperties, ParserOmissions>
//       : S extends 'parsed' ? Omit<PipelineProperties, ParsedOmissions>
//         : S extends 'sfcBlocksExtracted' ? Omit<PipelineProperties, SfcBlockOmissions>
//           : S extends 'closeout' ? PipelineProperties
//             : never

/**
 * The _state/payload_ that is available at a given stage in the pipeline process.
 *
 * - `<S>` provides the stage we're in
 */
export type Pipeline<S extends IPipelineStage> = S extends 'initialize'
  ? Omit<PipelineProperties, InitializedOmissions>
  : S extends 'metaExtracted' ? Omit<PipelineProperties, MetaOmissions>
    : S extends 'parser' ? Omit<PipelineProperties, ParserOmissions>
      : S extends 'parsed' ? Omit<PipelineProperties, ParsedOmissions>
        : S extends 'sfcBlocksExtracted' ? Omit<PipelineProperties, SfcBlockOmissions>
          : S extends 'closeout' ? PipelineProperties
            : never

/**
 * The Builder's event listener/handler
 */
export type BuilderHandler<
  O extends BuilderOptions,
  S extends IPipelineStage,
> = (payload: Pipeline<S>, options: O) => Promise<Pipeline<S>>

/**
 * Users configure a `BuilderHandler` and we wrap this up functionality
 * with a higher-order `BuilderTask`.
 *
 * @returns TaskEither<string, Pipeline<S>>
 */
export type BuilderTask<
  S extends IPipelineStage,
> = () => (payload: Pipeline<S>) => TE.TaskEither<string, Pipeline<S>>

/**
 * Builder's must provide an export which meets this API constraint. Basic
 * structure of this higher order function is:
 *
 * - options( ) -> register( ) -> { handler( payload ) -> payload }
 */
export type BuilderApi<O extends {}, S extends IPipelineStage> = (options?: O) => () => BuilderRegistration<O, S>

export type InlineBuilder = <N extends string, L extends IPipelineStage>(name: N, lifecycle: L) => (payload: Pipeline<L>) => Pipeline<L>

/**
 * Builder options are expected to be a key/value dictionary but must
 * be allowed to be an empty object
 * */
export type BuilderOptions = Record<string, any> | {}

export type BuilderConfig = Record<IPipelineStage, BuilderRegistration<any, any>[] | []>

/**
 * Carries an Either<T> condition which is either:
 * - a _string_ error condition
 * - a `Pipeline<S>` value
 */
export type PipeEither<S extends IPipelineStage> = Either<string, Pipeline<S>>

/**
 * Carries an `TaskEither<T>` condition which is either:
 * - a _string_ error condition
 * - a `() => Promise<Pipeline<S>>` task
 */
export type PipeTask<S extends IPipelineStage> = TE.TaskEither<string, Pipeline<S>>

/**
 * A pipeline payload or either an async `PipeTask<S>` or a synchronous `PipeEither<S>`
 */
export type PipelinePayload<S extends IPipelineStage> = PipeTask<S> | PipeEither<S>

/**
 * A _synchronous_ transformer function which:
 *
 * - recieves a payload of `PipeEither<F>`, and
 * - converts it to a type of `PipeEither<T>`
 */
export type SyncPipelineTransformer<
  F extends IPipelineStage,
  T extends IPipelineStage,
> = (payload: PipeTask<F>) => PipeTask<T>

/**
* An _asynchronous_ transformer function which:
*
* - recieves a payload of `PipeTask<F>` (async) or `PipeEither<F>` (sync)
* - converts it to a type of `PipeTask<T>`
*/
export type AsyncPipelineTransformer<
  F extends IPipelineStage,
  T extends IPipelineStage,
> = (payload: PipeTask<F>) => PipeTask<T>
