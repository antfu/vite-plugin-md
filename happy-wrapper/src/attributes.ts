import { string } from 'fp-ts'
import { pipe } from 'fp-ts/lib/function'
import type { IElement, INode } from 'happy-dom'
import { createFragment } from './create'
import { HappyMishap } from './errors'
import type { Container, DocRoot, GetAttribute, HTML } from './happy-types'
import { isDocument, isElement, isFragment } from './type-guards'
import { getNodeType, solveForNodeType, toHtml } from './utils'

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type SetAttribute<T extends string> = (value: string) => <N extends Container | HTML>(node: N) => N

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type SetAttributeTo<_T extends string, _V extends string> = <N extends Container | HTML>(node: N) => N

export const setAttribute = <T extends string>(
  attr: T,
): SetAttribute<T> => <V extends string>(
    value: V,
  ): SetAttributeTo<T, V> => <N extends Container | HTML>(
      node: N,
    ): N => {
    const invalidNode = (n: INode) => {
      throw new HappyMishap(`You can not use the setAttribute() utility on a node of type: "${getNodeType(n)}"`, { name: `setAttribute(${attr})(${value})(INVALID)` })
    }
    const result = solveForNodeType()
      .mirror()
      .solver({
        html: h => pipe(h, createFragment, f => setAttribute(attr)(value)(f), toHtml),
        text: t => invalidNode(t),
        node: n => invalidNode(n),
        fragment: (f) => {
          f.firstElementChild.setAttribute(attr, value)
          return f
        },
        document: (d) => {
          d.body.firstElementChild.setAttribute(attr, value)
          return d
        },
        element: (e) => {
          e.setAttribute(attr, value)
          return e
        },
      })(node)

    return result
  }

export const getAttribute = <T extends string>(attr: T): GetAttribute<T> => {
  return solveForNodeType('text', 'node')
    .outputType<string>()
    .solver({
      html: h => pipe(h, createFragment, getAttribute(attr)),
      fragment: f => f.firstElementChild.getAttribute(attr),
      document: doc => doc.body.firstElementChild.getAttribute(attr),
      element: el => el.getAttribute(attr),
    })
}

const getClass = getAttribute('class')
const setClass = setAttribute('class')
/**
 * Provides the classes defined on a given container's top level
 * element as an array of strings
 */
export const getClassList = (container: Container | HTML | null): string[] => {
  if (!container)
    return []

  return solveForNodeType().outputType<string[]>().solver({
    html: h => pipe(h, createFragment, getClassList),
    document: d => getClass(d.body.firstElementChild)?.split(/\s+/) || [],
    fragment: f => getClass(f.firstElementChild)?.split(/\s+/) || [],
    element: e => getClass(e)?.split(/\s+/) || [],
    text: (n) => { throw new HappyMishap('Passed in a text node to getClassList!', { name: 'getClassList', inspect: n }) },
    node: (n) => { throw new HappyMishap('Passed in an unknown node type to getClassList!', { name: 'getClassList', inspect: n }) },
  })(container).filter(i => i)
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type RemoveClass<R extends string | string[]> = <C extends DocRoot | IElement | HTML>(container: C) => C

/**
 * Removes a class from the top level node of a container's body.
 *
 * Note: if the class wasn't present then no change is performed
 */
export const removeClass = <R extends string | string[]>(
  remove: R,
): RemoveClass<R> => <D extends DocRoot | IElement | HTML>(doc: D): D => {
  const current = getClass(doc)?.split(/\s+/g) || []
  const toRemove: string[] = !Array.isArray(remove) ? [remove] : remove

  const resultantClassString = Array.from(
    new Set<string>(current.filter(c => !toRemove.includes(c))),
  )
    .filter(i => i)
    .join(' ')

  return setClass(resultantClassString)(doc)
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type AddClass<A extends string[] | string[][]> = <C extends DocRoot | IElement | HTML>(container: C) => C

/**
 * Adds a class to the top level node of a document's body.
 */
export const addClass = <A extends string[] | string[][]>(
  ...add: A
): AddClass<A> => <D extends DocRoot | IElement | HTML>(doc: D): D => {
  const toAdd = (Array.isArray(add) ? add.flat() : [add]) as string[]

  const currentClasses = getClass(doc)?.split(/\s+/g) || []
  const resultantClasses = Array.from(new Set<string>([...currentClasses, ...toAdd]))

  return setClass(resultantClasses.join(' ').trim())(doc) as D
}

export type Filter = (string | RegExp)
export type FilterCallback = (removed: string[]) => void
export type FiltersWithCallback = [FilterCallback, ...Filter[]]

function hasFilterCallback(filters: Filter[] | FiltersWithCallback): filters is FiltersWithCallback {
  return typeof filters[0] === 'function'
}

/**
 * Filters classes out from a given element using _filters_, where a filter:
 *
 * - string - when a string it will compare for a direct match
 * - RegExp - will run the RegExp's `test(class)` method
 *
 * Optionally you may pass in a callback function as the first parameter in the
 * the list and this callback will be then called with all filtered properties
 * passed to it. This is useful for creating desirable side-effects like _moving_
 * the classes to some other DOM element (for instance).
 */
export const filterClasses = <A extends Filter[] | [FilterCallback, ...Filter[]]>(
  ...args: A
) => <D extends DocRoot | IElement | HTML>(
    doc: D,
  ): D => {
  const el = isDocument(doc) || isFragment(doc)
    ? doc.firstElementChild as IElement
    : isElement(doc)
      ? doc as IElement
      : null
  if (!el)
    throw new HappyMishap('An invalid container was passed into filterClasses()!', { name: 'filterClasses', inspect: doc })

  const filters = hasFilterCallback(args) ? args.slice(1) as Filter[] : args as Filter[]
  const cb = hasFilterCallback(args) ? args[0] : undefined
  const classes = getClassList(el)
  const removed: string[] = []

  classes.forEach((klass) => {
    const matched = !filters.every(f =>
      typeof f === 'string'
        ? f.trim() !== klass.trim()
        : !f.test(klass),
    )
    if (matched)
      removed.push(klass)
  })

  setClass(classes.filter(k => !removed.includes(k)).join(' '))(doc)

  if (cb)
    cb(removed)

  return doc
}
