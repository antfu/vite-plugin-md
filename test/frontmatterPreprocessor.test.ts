import { readFile } from 'fs/promises'
import { beforeAll, describe, expect, it } from 'vitest'
import { composeSfcBlocks } from '../src/pipeline'
import type { MetaProperty, ResolvedOptions } from '../src/types'

const frontmatterPreprocess: ResolvedOptions['frontmatterPreprocess'] = (fm) => {
  const frontmatter = {
    title: 'default title',
    description: 'default description',
    ...fm,
  }
  const meta: MetaProperty[] = [
    { property: 'og:title', name: 'twitter:title', itemprop: 'title', content: frontmatter.title },
    {
      property: 'og:description',
      name: 'twitter:description',
      itemprop: 'description',
      content: frontmatter.description,
    },
  ]
  return {
    head: { ...frontmatter, meta },
    frontmatter: { ...frontmatter, meta },
    metaProps: [],
    routeMeta: {},
  }
}

let md = ''

describe('frontmatter pre-processor', () => {
  beforeAll(async () => {
    md = await readFile('test/fixtures/simple.md', 'utf-8')
  })

  it('frontmatter is unchanged', async () => {
    const { frontmatter } = await composeSfcBlocks('', md, { frontmatterPreprocess, headEnabled: true })
    expect(frontmatter).toMatchSnapshot()
  })

  it('head is unchanged', async () => {
    const { head } = await composeSfcBlocks('', md, { frontmatterPreprocess, headEnabled: true })
    expect(head).toMatchSnapshot()
  })

  it('meta props are unchanged', async () => {
    const { meta } = await composeSfcBlocks('', md, { frontmatterPreprocess, headEnabled: true })
    expect(meta).toMatchSnapshot()
  })

  it('inline markdown is used over default properties', async () => {
    const { frontmatter } = await composeSfcBlocks('', md, { frontmatterPreprocess, headEnabled: true })

    // Positive tests
    expect(
      frontmatter.title?.includes('Hello World'),
      'the title attribute is retained over the default \'title\' value',
    ).toBeTruthy()

    expect(
      frontmatter.description?.includes('testing is the path to true happiness'),
      'description property is also retained',
    ).toBeTruthy()

    // Negative tests
    expect(
      frontmatter.title?.includes('default title'),
      'the title attribute is retained over the default \'title\' value',
    ).toBeFalsy()

    expect(
      frontmatter.description?.includes('default description'),
      'default description is ignored',
    ).toBeFalsy()
  })

  it('meta and head props are populated based on default rules', async () => {
    const { head, meta } = await composeSfcBlocks('', md, { frontmatterPreprocess, headEnabled: true })
    // Meta props
    const title = meta.find(i => i.property === 'og:title')
    const desc = meta.find(i => i.property === 'og:description')

    expect(head).toBeDefined()
    expect(head.title).toBeDefined()
    expect(head.meta).toBeDefined()

    expect(title).toBeDefined()
    expect(desc).toBeDefined()
    expect(title?.property).toEqual('og:title')
    expect(desc?.property).toEqual('og:description')
  })
})
