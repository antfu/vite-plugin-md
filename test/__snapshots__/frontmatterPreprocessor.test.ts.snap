// Vitest Snapshot v1

exports[`frontmatter pre-processor > frontmatter is unchanged 1`] = `undefined`;

exports[`frontmatter pre-processor > head is unchanged 1`] = `undefined`;

exports[`frontmatter pre-processor > meta props are unchanged 1`] = `undefined`;
